New sio backend

tmadlener opened the pull request:
BEGINRELEASENOTES
- Add SIO as a second I/O backend (as alternative to ROOT) that can be enabled with `ENABLE_SIO`. If enabled, a separate `podioSioIO` library is built that allows reading and writing sio files. For serializing the different datatypes, additional code is generated to build an `SioBlocks` library that is loaded at runtime (if found somewhere on `LD_LIBRARY_PATH`). To facilitate the whole process at the cmake level, **new cmake functions** are provided to generate the core datamodel library **`PODIO_ADD_DATAMODEL_CORE_LIBRARY`**, to (conditionally) define the ROOT dictionary target **`PODIO_ADD_ROOT_IO_DICT`** and to (conditionally) define the Sio Blocks library target **`PODIO_ADD_SIO_IO_BLOCKS`**. The I/O backends that are supported by podio are exported via the **`PODIO_IO_HANDLERS`** list variable.
- `podio_generate_datamodel.py` now additionally takes the I/O handlers that should be generated as arguments. This is also reflected in an additional argument to `PODIO_GENERATE_DATAMODEL`. To have backwards compatibility, this additional argument defaults to ROOT in both cases and downstream packages should work as usual without changes.

ENDRELEASENOTES

This is more or less a copy/paste exercise from a previous attempt from Frank with the main changes in the templates and the c++ code generator. At the moment this should be considered as a prototype version and there are still some things to do:

- [x] More detailed Release Notes
- [x] Cleanup code
- [x] Make `read_sio` test run successfully via `ctest`
- [x] Update CI to also have SIO available
- [x] I/O of meta data
  - [x] collection IDs (and other information necessary to reconstruct the collections after reading)
  - [x] event meta data
  - [x] run meta data
  - [x] collection meta data
- [x] separate SIO library part from core library part (similar to what is done for root)
- [x] export targets properly for downstream packages

gaede commented:
@tmadlener: you had commented out reading of the meta data in the first record, which is uncompressed (for whatever reason) and then reading the first event in compressed mode failed...

tmadlener commented:
I have added the handling of the event meta data in a first version and also slightly reworked the handling of the collection IDs and the data that is necessary to reconstruct all collections afterwards. I have temporarily disabled the tests that check the collection meta data, as that still needs to be done.

**Currently the changes in #133 and #134 are also present here**

tmadlener commented:
From a functionality point of view I think everything that is covered by the `ROOTReader` and 'ROOTWriter' is now also covered by the `SIOReader` and `SIOWriter`, with the exception of reading multiple files.

There are, however, a few things that could be improved in the current implementation:
- The `GenericParameters` now have public methods to access their internal maps, solely for the purpose of easier I/O using `sio`. I don't particularly like this.
- The `SIOBlockFactory` with its static instance, is not really ideal for multi threading. From the looks of it, it could potentially be replaced by a constant map that could be generated from the yaml file, but I am not entirely sure about that.
- The `SIOReader` currently owns the collections, which makes the `EventStore` behave slightly differently, as compared with when it is used with the `ROOTReader` (i.e. `EventStore::clear` will destroy resources that are not owned by it). Ideally the two readers would be drop-in replacable.

For these reasons I am currently keeping this as [WIP], but any intermediate comments are very welcome.

tmadlener commented:
I am currently not entirely happy with the splitting of the different libraries, especially for the tests, where I have tried to have a "core" library, as well as two separate libraries that are only concerned with either root or sio based I/O. For root I have basically followed the same path as in [edm4hep](https://github.com/key4hep/EDM4hep/blob/master/edm4hep/CMakeLists.txt) and this seems to work nicely out of the box. However, for sio I am currently struggling with getting the self registering factory approach to work nicely. The problem is that all the `SIOBlock`s that are necessary for this are placed in a separate library and so the `SIOBlockFactory` instance in the tests will not know about the one that is created and filled in this library. Including one `SIOBlock` header in the tests fixes this for now, but this is not really a solution in my opinion.

For the tests here, it doesn't really matter, but for actual usage it might be nice to separate the core libraries from the one that deal with I/O. Is there an easy way to do this?

andresailer commented on CMakeLists.txt:
```diff
@@ -79,6 +80,15 @@ if(ROOT_VERSION VERSION_LESS 6.14)
+  find_package( SIO REQUIRED)
+  if (SIO_FOUND)
+    MESSAGE( STATUS "Found SIO library - will build SIO I/O support" )
+    include_directories(${SIO_INCLUDE_DIRS})
```
Doesn't SIO export the targets?
In any case, don't use `include_directories`, use `target_include_directories` for the applicable targets

tmadlener replied:
No, unfortunately SIO does not (yet) export its targets.

andresailer commented on python/podio_class_generator.py:
```diff
@@ -151,7 +151,8 @@ def _get_filenames_templates(template_base, name):
         'Object': '',
-        'Component': ''
+        'Component': '',
+        'SIOBlock': 'SIOBlock'
```
```suggestion
        'SIOBlock': 'SIOBlock',
```
Avoids noisy diffs next time

andresailer commented on python/podio_class_generator.py:
```diff
@@ -205,6 +206,9 @@ def _process_datatype(self, name, definition):
     self._fill_templates('Obj', datatype)
     self._fill_templates('Collection', datatype)
 
+    if self.reader.options["createSIOHandlers"]:
```
```suggestion
    if 'SIO' in self.reader.options["IOHandlers"]:
```
Or something like this, and making IOHandlers a list of IOHandlers to create (ROOT and SIO for the moment)

tmadlener replied:
That is a good idea. It would definitely scale better than the current approach in the yaml file of the datamodel. I will make this change, and since ROOT doesn't need any special treatment at the moment it will basically be simply ignored. Do you think that is a reasonable approach, or should we do something more explicit for IO systems that do not need additional generated code? E.g. only list the handlers that actually need additional code generation?

andresailer replied:
After reading more of the PR.

Maybe this should be coupled to the CMake option for enabling SIO, in the base CMakeLists in the block
```cmake
if(SIO_HANDLERS)
  LIST(APPEND PODIO_IO_HANDLERS SIO)
endif()
```

The PODIO_GENERATE_DATAMODEL function could just set environment variables for each handler,
```cmake
FOREACH(HANDLER PODIO_IO_HANDLERS)
  SET(EXPORT_HANDLER "${EXPORT_HANDLER} ${HANDLER}_HANDLER=1")
ENDFOREACH()
```
This

https://github.com/AIDASoft/podio/blob/d4a02651332e2fe3ffbac4e99fa61b34600a16fa/cmake/podioMacros.cmake#L141

becomes
```
    COMMAND ${EXPORT_HANDLER} python ${podio_PYTHON_DIR}/podio_class_generator.py --quiet ${YAML_FILE} ${ARG_OUTPUT_FOLDER} ${datamodel}
```
and then the check becomes
```python
if os.environ.get('SIO_HANDLER'):
```

The variable  PODIO_IO_HANDLERS is exported to the podioConfig, so edm4hep can pick this up as well when calling PODIO_GENERATE_DATAMODEL

Which reminds me that the dependency on SIO should also be exported in the podioConfig, if enabled.

tmadlener replied:
> ROOT doesn't need any special treatment 

Actually, I just realized that this is not true, because the `selection.xml` file is only necessary for the ROOT dictionaries.

-----------
Regarding the handling of the generation of different IO systems, I have started some kind of discussion in that direction in #131. Maybe instead of setting environment variables, we could pass command line arguments to the generator from cmake? Or would that defeat the purpose of exporting `PODIO_IO_HANDLERS`? I, personally, find explicitly passing command line arguments a bit easier to debug and to follow the flow of the whole thing.
 
For exporting the dependency on SIO, it would be ideal if SIO already exports targets, right?

andresailer commented on python/templates/SIOBlock.cc.jinja2:
```diff
@@ -0,0 +1,90 @@
+#include "{{ incfolder }}{{ class.bare_type }}SIOBlock.h"
+
+#include <sio/block.h>
+#include <sio/version.h>
```
sort includes alphabetically?

tmadlener replied:
Done. Also in the header templates.

andresailer commented on src/CMakeLists.txt:
```diff
@@ -48,10 +56,27 @@ set_target_properties(podioDict-dictgen PROPERTIES EXCLUDE_FROM_ALL TRUE)
-  DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+# SIO I/O library
+if(SIO_HANDLERS)
+add_library(podioSioIO SHARED ${sio_sources})
```
Needs indentation

andresailer commented on src/CMakeLists.txt:
```diff
@@ -48,10 +56,27 @@ set_target_properties(podioDict-dictgen PROPERTIES EXCLUDE_FROM_ALL TRUE)
+  $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+target_link_libraries(podioSioIO PUBLIC podio::podio ${SIO_LIBRARIES})
+
```
```suggestion
  LIST(APPEND INSTALL_LIBRARIES podioSioIO)
```

andresailer commented on src/CMakeLists.txt:
```diff
@@ -48,10 +56,27 @@ set_target_properties(podioDict-dictgen PROPERTIES EXCLUDE_FROM_ALL TRUE)
+    DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+else()
+  # Install the Targets and Headers
+  install(TARGETS podio podioDict podioRootIO
```
```suggestion
  install(TARGETS podio podioDict podioRootIO ${INSTALL_LIBRARIES}
```
Then you don't need different install commands, if the variable is not set it isn't a problem, but it scales to more IOHandlers easily.
I hope a list works here.

tmadlener replied:
Just for future reference: a list seems to work nicely here.

andresailer commented on src/CMakeLists.txt:
```diff
@@ -48,10 +56,27 @@ set_target_properties(podioDict-dictgen PROPERTIES EXCLUDE_FROM_ALL TRUE)
+if(SIO_HANDLERS)
+add_library(podioSioIO SHARED ${sio_sources})
+add_library(podio::podioSioIO ALIAS podioSioIO)
+target_include_directories(podio PUBLIC
```
```suggestion
target_include_directories(podioSioIO PUBLIC
```
Right?

andresailer commented on tests/CMakeLists.txt:
```diff
@@ -7,52 +7,91 @@ endforeach()
+target_include_directories(TestDataModelDict PUBLIC
+  ${CMAKE_SOURCE_DIR}/include
+  ${CMAKE_CURRENT_SOURCE_DIR}/
+  ${ROOT_INCLUDE_DIRS}
```
```suggestion
```
Must not be needed because linking ROOT::Core

andresailer commented on src/SIOReader.cc:
```diff
@@ -0,0 +1,148 @@
+#include "sio/definitions.h"
+
+#include <memory>
+#include <sio/compression/zlib.h>
```
Does this mean SIO has to be build with the builtin zlib? Because similar things have caused problems in the past (https://github.com/AIDASoft/DD4hep/issues/567)

tmadlener replied:
No, SIO can be built with its builtin zlib, but it can also use the system one. For actual usage it should not matter which zlib is actually used in the end. I have briefly tested this with rebuilding SIO with the system zlib, and for podio the tests still work.

andresailer replied:
Ok, thanks for checking.

tmadlener commented:
With the latest changes I have now implemented the following:
- A very basic runtime library loader in form of the `SIOBlockLibraryLoader`. It is designed to be used as a singleton that tries to load the necessary libraries during construction. It is then simply instantiated in the constructors of the `SIOReader` and `SIOWriter`. To load libraries all directories on `LD_LIBRARY_PATH` are checked for files containing `"SioBlocks"` in their name, which are then loaded. Since this also implies a naming convention, this does the job for the moment, but maybe this should be made more stable in the future.
- `SIOReader` now follows the same interface as `ROOTReader` and should now be in principle usable as a drop-in replacement. However, some differences still remain due to the different philosophies of SIO and ROOT, e.g. ROOT offers a `getEntries` method, while SIO signals the end of the file with an exception and a corresponding error code.
- Replaced the `createSIOHandlers` option with an `IOHandlers` one for configuring which IO handler code should be generated. I would put a potential switch to command line arguments (see discussion [above](https://github.com/AIDASoft/podio/pull/130#discussion_r488731259) and in #131) in a different PR to not pollute this one too much.
- Addressed Andres `cmake` comments

hegner commented on CMakeLists.txt:
```diff
@@ -68,6 +68,7 @@ endif()
 
 #--- Declare options -----------------------------------------------------------
 option(CREATE_DOC "Whether or not to create doxygen doc target." OFF)
+option(SIO_HANDLERS "Build the SIO block handlers and readers and writers." OFF)
```
naming a bit too technical. Why not "SIO_SUPPORT" or "ENABLE_SIO" ? One can use equivalent naming for every further backend.

tmadlener replied:
I have changed it to `ENABLE_SIO`. I realized that this has to be done all over the place, so we should come up with a more "compact" way of adding backends.

hegner replied:
yap. that was my comment in the meeting already. thanks!

tmadlener replied:
I think we can broaden the discussion in #131 to cover the whole topic of how to best deal with different I/O backends.

hegner commented on python/podio_config_reader.py:
```diff
@@ -321,6 +321,8 @@ def __init__(self, yamlfile):
         "exposePODMembers": True,
         # use subfolder when including package header files
         "includeSubfolder": False,
+        # generate the SIO block handlers
```
certainly comment and code are out of sync ;-)

tmadlener replied:
Ah yes, most definitely... Thanks for checking

andresailer commented on tests/CMakeLists.txt:
```diff
@@ -7,52 +7,91 @@ endforeach()
-file(GLOB sources_utils utilities/*.cc)
-file(GLOB headers_utils utilities/*.h)
+if (ENABLE_SIO)
+  # Filter out the sio block handlers to avoid polluting the TestDataModel lib
```
The same issue will appear in edm4hep.
Can we find a way to make this automated for users of PODIO_GENERATE_DATAMODEL?

tmadlener replied:
Yes, we definitely should. I would suggest to discuss this in #131.

andresailer commented on tests/CMakeLists.txt:
```diff
@@ -7,52 +7,91 @@ endforeach()
+set(root_libs TestDataModelDict podio::podioRootIO)
+set(root_env ROOT_INCLUDE_PATH=${CMAKE_SOURCE_DIR}/tests/datamodel:${ROOT_INCLUDE_PATH})
+foreach( sourcefile ${root_dependent_tests} )
+  CREATE_PODIO_TEST(${sourcefile} "${root_libs}" ${root_env})
```
```suggestion
  CREATE_PODIO_TEST(${sourcefile} "${root_libs}")
```
Given the signature of CREATE_PODIO_TEST `root_env` isn't needed, or `additional_env` is missing there.

(I was wondering if it should be `$ENV{ROOT_INCLUDE_PATH}` in the line above, but if it is not used it doesn't matter.)

tmadlener replied:
Another remnant from my intermediate changes. Originally I think I had `additional_env` as argument to `CREATE_PODIO_TEST`, just for passing in the `ROOT_INCLUDE_PATH` but since that is not necessary, I will completely remove all traces of it, also for the other tests.

hegner commented on tests/read_test.h:
```diff
@@ -0,0 +1,226 @@
+    throw std::runtime_error("Couldn't read event meta data parameters 'UserEventWeight'");
+  }
+  std::stringstream ss ; ss << " event_number_" << eventNum ;
+  std::string evtName = evtMD.getStringVal( "UserEventName" ) ;
```
If you would have used *const auto*, you would have gotten *const std::string&* instead of a copy.

tmadlener replied:
Ah yes, this is true. But these changes are currently only here because this PR starts on top of #133. I will fix it there. There are probably also a few other instances, where the usage of `auto` would make things a bit nicer in the tests. I haven't really checked, as I have basically just moved the already available code into a common header.

gaede commented:
@tmadlener nice work. I have pushed an option to use the boost_filesystem library in SioIO as filesystem is not yet available on all Macs (at least on my mac running Mojave).

andresailer commented:
@gaede  Have you tried linking against c++fs with llvm, instead of bringing in boost?
See:
https://github.com/AIDASoft/DD4hep/blob/a2f3c03d74f27e24fa8b4a1b8b7478ae654d8a35/cmake/DD4hepBuild.cmake#L721
Or is dd4hep falling back on boost on your machine?

andresailer commented:
I think we need something like https://github.com/AIDASoft/DD4hep/blob/a2f3c03d74f27e24fa8b4a1b8b7478ae654d8a35/cmake/DD4hepBuild.cmake#L721-L751 to support compilers older than gcc9.1

tmadlener commented:
Hi @andresailer,
can you check whether I got the essential parts right for detecting a suitable filesystem library in https://github.com/AIDASoft/podio/pull/130/commits/74a9486da5d4ac7935d3a22a4f39c87697106649 ?

@gaede, I added a simple TOC in https://github.com/AIDASoft/podio/pull/130/commits/c624808ced3e1c46ed7da816b39cd2330166f3fb as we have discussed. I am not yet entirely sure whether placing a marker outside of sio blocks at the end of the file really plays nicely with everything, but the tests don't uncover anything unexpected at the moment.

andresailer commented on cmake/podioMacros.cmake:
```diff
@@ -151,3 +151,40 @@ function(PODIO_GENERATE_DATAMODEL datamodel YAML_FILE RETURN_HEADERS RETURN_SOUR
+
+#---------------------------------------------------------------------------------------------------
+function(PODIO_CHECK_CPP_FS FS_LIBS)
+  SET(${have_filesystem} FALSE)
```
```suggestion
  SET(have_filesystem FALSE)
```

andresailer commented on cmake/podioMacros.cmake:
```diff
@@ -151,3 +151,40 @@ function(PODIO_GENERATE_DATAMODEL datamodel YAML_FILE RETURN_HEADERS RETURN_SOUR
+    IF(have_filesystem)
+      MESSAGE(VERBOSE "Compiler supports filesystem when linking against ${FS_LIB_NAME}")
+      SET(${FS_LIBS} ${FS_LIB_NAME} PARENT_SCOPE)
+      BREAK()
```
```suggestion
      RETURN()
```
As this is a function, we can use RETURN(). It makes no real difference, but we could drop the `IF(NOT have_filesystem)`

andresailer commented on cmake/podioMacros.cmake:
```diff
@@ -151,3 +151,40 @@ function(PODIO_GENERATE_DATAMODEL datamodel YAML_FILE RETURN_HEADERS RETURN_SOUR
+      PROPERTIES
+      INTERFACE_COMPILE_DEFINITIONS USE_BOOST_FILESYSTEM
+    )
+  GET_TARGET_PROPERTY(BOOST_FILESYSTEM_LOC Boost::filesystem IMPORTED_LOCATION)
```
These two lines don't do anything in this case. I am not even sure of they do anything in the DD4hep macro case.

tmadlener replied:
Ah, I see. I didn't realize that. I have removed them.

andresailer reviewed (commented):
Comments for the filesystem checks

tmadlener commented on cmake/podioMacros.cmake:
```diff
@@ -151,3 +151,40 @@ function(PODIO_GENERATE_DATAMODEL datamodel YAML_FILE RETURN_HEADERS RETURN_SOUR
+#---------------------------------------------------------------------------------------------------
+function(PODIO_CHECK_CPP_FS FS_LIBS)
+  SET(${have_filesystem} FALSE)
+  MESSAGE(VERBOSE "Checking for filesystem support of compiler")
```
ah yes, I have apparently not checked the documentation of older versions enough. It becomes available with cmake > 3.15. I have removed some of the output and made some other messages to be `STATUS` level.

petricm replied:
this triggered that the PYTHON check in #141 needs cmake 3.12 because FindPython exists only since that version.

andresailer replied:
Ahh, nice, good to know for the future :+1:

tmadlener commented:
I have added some cmake functions to facilitate the downstream usage of the different I/O backends. Additionally I have added `podio_IO_HANDLERS` to the exported variables. The cmake functions do not provide a complete automatization of the downstream process, but should facilitate the usage. Example usage can be seen in the tests directory. I have briefly checked with edm4hep to see if things actually work downstream. With the usage of these new macros the [edm4hep/CMakeLists.txt](https://github.com/key4hep/EDM4hep/blob/master/edm4hep/CMakeLists.txt) could look something like this:

```cmake
# For now unconditionally generate all the code that is supported by the intalled podio
PODIO_GENERATE_DATAMODEL(edm4hep ../edm4hep.yaml headers sources IO_BACKEND_HANDLERS ${podio_IO_HANDLERS})

PODIO_ADD_DATAMODEL_CORE_LIB(edm4hep "${headers}" "${sources}")

PODIO_ADD_ROOT_IO_DICT(edm4hepDict edm4hep "${headers}" src/selection.xml)
IF(TARGET edm4hepDict)
  add_library(edm4hep::edm4hepDict ALIAS edm4hepDict )
ENDIF()

PODIO_ADD_SIO_IO_BLOCKS(edm4hep "${headers}" "${sources}")
IF(TARGET edm4hepSioBlocks)
  message(STATUS "SioBlocks are here!")
ENDIF()

# 
# The rest: Installing targets, etc..
# 
```

Depending on whether podio actually provides capabilities to do I/O via SIO edm4hep would also build the corresponding library. It is still necessary to check whether it is actually available, but the boilerplate code to conditionally add it is hidden in the cmake function.

It is also still possible to pass different `IO_BACKEND_HANDLERS` to `PODIO_GENERATE_DATAMODEL` and the functions still work, because they only check for generated files and act depending on whether it is present or not.

tmadlener commented:
I have updated the release notes and I think this PR is now ready for (another) review. I am keeping at as [WIP] at the moment, because it still contains the changes from #133, and I think it would be best if that is merged first. Additionally, also #141 could be merged first to see if everything still works with the udpated CI.

After that it might be necessary to rebase. In the course of that also the final review comments can be addressed and then it should be possible to merge this.

andresailer commented on cmake/podioMacros.cmake:
```diff
@@ -138,8 +146,170 @@ function(PODIO_GENERATE_DATAMODEL datamodel YAML_FILE RETURN_HEADERS RETURN_SOUR
+
+
+#---------------------------------------------------------------------------------------------------
+#---PODIO_ADD_ROOT_IO_LIBS( dict_name CORE_LIB HEADERS SELECTION_XML
```
```suggestion
#---PODIO_ADD_ROOT_IO_DICT( dict_name CORE_LIB HEADERS SELECTION_XML
```

tmadlener replied:
Very good catch. Thanks.

tmadlener commented:
Removed [WIP] and rebased on top of master. Barring any further review comments, this is ready to be merged from my side.

hegner commented on CMakeLists.txt:
```diff
@@ -68,6 +68,8 @@ endif()
 
 #--- Declare options -----------------------------------------------------------
 option(CREATE_DOC "Whether or not to create doxygen doc target." OFF)
+option(ENABLE_SIO "Build the SIO block handlers and readers and writers." OFF)
```
Less technical: "Build SIO read/write backend"

hegner replied:
or like said further below "build SIO I/O support"

hegner commented:
Thanks @tmadlener !

I will review it during the day. It is quite a long one.

andresailer commented on cmake/podioConfig.cmake.in:
```diff
@@ -21,6 +21,7 @@
+set(podio_IO_HANDLERS "@PODIO_IO_HANDLERS@")
 
 include(CMakeFindDependencyMacro)
 find_dependency(ROOT @ROOT_VERSION@)
```
There needs to be something for  `find_dependency(SIO)` here, if enabled, no?

tmadlener replied:
I am actually not entirely sure if that is necessary. In the spirit of "it works on my machine" I have successfully built `edm4hep` with SIO support and also been able to read/write files without adding this.

Could it be that this is propagated through one of the targets that links to `SIO` by any chance? Like `podioSioIO` for example? Or is it enough that all the places where we would actually require to link to `SIO` are hidden inside cmake macros?

In the end it probably won't hurt to include it in any case, I suppose?

andresailer replied:
Yes, at the moment the targets probably contain the full path to the SIO library and include directories used at build time. But if SIO moves to targets this would start failing.

tmadlener replied:
That makes sense. I have added this in the last commit.

tmadlener commented on cmake/podioConfig.cmake.in:
```diff
@@ -21,6 +21,7 @@
 set_and_check(podio_INCLUDE_DIR "@PACKAGE_CMAKE_INSTALL_INCLUDEDIR@")
 set_and_check(podio_LIBRARY_DIR "@PACKAGE_CMAKE_INSTALL_LIBDIR@")
 set_and_check(podio_PYTHON_DIR "@PACKAGE_podio_PYTHON_INSTALLDIR@")
+set(podio_IO_HANDLERS "@PODIO_IO_HANDLERS@")
```
I was aiming for consistency with the other exported variables, where `podio` is used in lower case ;)
But it would indeed be more consistent in upper case, also with regards to the scheme for the macros.

tmadlener replied:
Changed, and also updated the release notes.

gaede commented:
> Thanks @tmadlener !
> 
> I will review it during the day. It is quite a long one.

Hi @hegner, did you find the time to review/approve this PR ? We'd like to merge it soon...

tmadlener commented:
Closing and opening again, to retrigger CI.

tmadlener commented:
close + reopen to trigger CI
